op-service/eth/blob.go

```diff
@@ -70,60 +79,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+		return nil, fmt.Errorf("%w: expected version %d, got %d", InvalidEncodingVersionError, EncodingVersion, b[VersionOffset])
+	}
+
+	// decode the 3-byte big-endian length value into a 4-byte integer
+	outputLen := uint32(b[2])<<16 | uint32(b[3])<<8 | uint32(b[4])
+	if outputLen > MaxBlobDataSize {
+		return nil, fmt.Errorf("%w: got %d", InvalidLengthError, outputLen)
+	}
+
+	// round 0 is special cased to copy only the remaining 27 bytes of the first field element into
+	// the output due to version/length encoding already occupying its first 5 bytes.
+	output := make(Data, MaxBlobDataSize)
+	copy(output[0:27], b[5:])
+
+	// now process remaining 3 field elements to complete round 0
+	opos := 28 // current position into output buffer
+	ipos := 32 // current position into the input blob
+	var err error
+	encodedByte := make([]byte, 4) // buffer for the 4 6-bit chunks
+	encodedByte[0] = b[0]
+	for i := 1; i < 4; i++ {
+		encodedByte[i] = b[ipos]
+		if opos, ipos, err = b.decodeFieldElement(opos, ipos, encodedByte, output); err != nil {
+			return nil, err
+		}
+	}
+	opos = reassembleBytes(opos, encodedByte, output)
+
+	// in each remaining round we decode 4 field elements (128 bytes) of the input into 127 bytes
+	// of output
+	for i := 1; i < Rounds && opos < int(outputLen); i++ {
+		for j := 0; j < 4; j++ {
+			// save the first byte of each field element for later re-assembly
+			encodedByte[j] = b[ipos]
+			if opos, ipos, err = b.decodeFieldElement(opos, ipos, encodedByte, output); err != nil {
+				return nil, err
+			}
 		}
-		copy(data[i*31:i*31+31], b[i*32+1:i*32+32])
+		opos = reassembleBytes(opos, encodedByte, output)
 	}
-	// extract the length prefix & trim the output accordingly
-	dataLen := binary.LittleEndian.Uint32(data[:4])
-	data = data[4:]
-	if dataLen > uint32(len(data)) {
-		return nil, fmt.Errorf("invalid blob, length prefix out of range: %d", dataLen)
+	output = output[:outputLen]
+	return output, nil
+}
+
+// decodeFieldElement decodes the next input field element by writing its lower 31 bytes into its
+// appropriate place in the output and checking the high order byte is valid. Returns an
+// InvalidFieldElementError if a field element is seen with either of its two high order bits set.
+func (b *Blob) decodeFieldElement(opos, ipos int, encodedByte []byte, output []byte) (int, int, error) {
+	// two highest order bits of the first byte of each field element should always be 0
+	if b[ipos]&(0xC0) != 0 {
+		return 0, 0, fmt.Errorf("%w: field element: %d", InvalidFieldElementError, ipos)
 	}
-	data = data[:dataLen]
-	return data, nil
+	copy(output[opos:], b[ipos+1:ipos+32])
+	return opos + 32, ipos + 32, nil
+}
+
+// reassembleBytes takes the 4x6-bit chunks from encodedByte, reassembles them into 3 bytes of
+// output, and places them in their appropriate output positions.
+func reassembleBytes(opos int, encodedByte []byte, output []byte) int {
+	opos-- // account for fact that we don't output a 128th byte
```

Comment:
Should we be worried about underflow here?

Reply:
reassembleBytes is only called after processing 4 field elements which advances opos by 128.  Max lookback here is within that window, so unless someone introduces a coding error or starts calling this from elsewhere, underflow can't happen.  Could extend this func to also return error if you want me to add an explicit test just in case.